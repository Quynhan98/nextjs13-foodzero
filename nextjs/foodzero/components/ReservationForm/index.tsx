import { ChangeEvent, FormEvent, memo } from 'react'
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormErrorMessage,
  Heading,
  Input,
  Text,
} from '@chakra-ui/react'

// Components
import Picker from '@components/Picker'
import SelectField from '@components/SelectField'

// Constants
import { NUMBER_OF_PERSON, RESERVATION_TIME } from '@constants/index'

interface ReservationFormProps {
  onSubmitForm: (e: FormEvent<HTMLFormElement>) => void
  onChangeField?: (e: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => void
  handleChangeDate: (date: Date) => void
  isShowFullField?: boolean
  isDisableButton?: boolean
  isDisableField?: boolean
  phoneError?: string
  emailError?: string
  timeError?: string
  date?: string
  phone?: string
  firstName?: string
  lastName?: string
  email?: string
  time?: string
  person?: string
}

const ReservationForm = ({
  onSubmitForm,
  onChangeField,
  handleChangeDate,
  isShowFullField = false,
  isDisableButton,
  isDisableField,
  phone,
  phoneError,
  emailError,
  timeError,
  date,
  firstName,
  lastName,
  email,
  time,
  person,
}: ReservationFormProps) => {
  return (
    <form onSubmit={onSubmitForm} noValidate>
      <FormControl
        isInvalid={!!(emailError || phoneError || timeError)}
        display="flex"
        flexDirection="column"
        justifyContent="center"
        maxW={{ base: '320px', md: '1148px', '2xl': '1644px' }}
        margin="0 auto"
      >
        <Flex flexDirection="column" justifyContent="center" margin="0 auto">
          <Heading
            as="h3"
            size="large"
            variant="secondary"
            pb={{ base: '10px', '2xl': '18px' }}
          >
            Make a Reservation
          </Heading>
          <Text
            pb={{ base: '50px', '2xl': isShowFullField ? '115px' : '133px' }}
            textAlign={isShowFullField ? 'start' : 'center'}
            size="default"
          >
            Get in touch with restaurant
          </Text>
        </Flex>

        {isShowFullField && (
          <>
            <Flex
              flexDirection={{ base: 'column', '2xl': 'row' }}
              gap={{ base: '0px', '2xl': '48px' }}
            >
              <Input
                value={firstName}
                disabled={isDisableField}
                onChange={onChangeField}
                name="firstName"
                type="text"
                placeholder="First Name"
                mb={{ base: '18px', '2xl': '48px' }}
              />
              <Input
                value={lastName}
                disabled={isDisableField}
                onChange={onChangeField}
                name="lastName"
                type="text"
                placeholder="Last Name"
                mb={{ base: '18px', '2xl': '48px' }}
              />
            </Flex>
            <Box mb={{ base: '18px', '2xl': '48px' }}>
              <Input
                value={email}
                disabled={isDisableField}
                onChange={onChangeField}
                name="email"
                type="email"
                placeholder="Email"
              />
              {emailError && (
                <FormErrorMessage fontSize={{ base: 'xxs', '2xl': 'base' }}>
                  {emailError}
                </FormErrorMessage>
              )}
            </Box>
            <Box mb={{ base: '18px', '2xl': '48px' }}>
              <Input
                value={phone}
                disabled={isDisableField}
                onChange={onChangeField}
                name="phone"
                type="tel"
                placeholder="Phone"
              />
              {phoneError && (
                <FormErrorMessage fontSize={{ base: 'xxs', '2xl': 'base' }}>
                  {phoneError}
                </FormErrorMessage>
              )}
            </Box>
          </>
        )}
        <Flex
          flexDirection={{
            base: 'column',
            md: isShowFullField ? 'column' : 'row',
          }}
          gap={{ base: '18px', md: '48px' }}
        >
          <Flex
            justifyContent="space-between"
            flexDirection={{ base: 'column', md: 'unset' }}
            gap={{ base: '18px', md: '48px' }}
          >
            <Box>
              <Picker
                pickerValue={date}
                disabled={isDisableField}
                width={{
                  base: '320px',
                  md: isShowFullField ? '550px' : '350px',
                  '2xl': isShowFullField ? '798px' : '516px',
                }}
                onChangeDate={handleChangeDate}
                isMinDate
                isMaxDate
              />
            </Box>
            <Box>
              <SelectField
                name="time"
                value={time}
                placeholder={time}
                isDisabled={isDisableField}
                data-testid="selectTime"
                options={RESERVATION_TIME}
                onChangeSelect={onChangeField}
                width={{
                  base: '100%',
                  md: isShowFullField ? '550px' : '350px',
                  '2xl': isShowFullField ? '798px' : '516px',
                }}
              />
              {timeError && (
                <FormErrorMessage fontSize={{ base: 'xxs', '2xl': 'base' }}>
                  {timeError}
                </FormErrorMessage>
              )}
            </Box>
          </Flex>
          <SelectField
            name="person"
            value={person}
            isDisabled={isDisableField}
            data-testid="selectPerson"
            placeholder={person}
            options={NUMBER_OF_PERSON}
            onChangeSelect={onChangeField}
          />
        </Flex>
        <Box
          margin="0 auto"
          pt={{ base: '50px', '2xl': isShowFullField ? '80px' : '133px' }}
        >
          <Button
            type="submit"
            variant={isShowFullField ? 'default' : 'dark'}
            size="primary"
            isDisabled={isDisableButton}
          >
            Book Now
          </Button>
        </Box>
      </FormControl>
    </form>
  )
}

export default memo(ReservationForm)
