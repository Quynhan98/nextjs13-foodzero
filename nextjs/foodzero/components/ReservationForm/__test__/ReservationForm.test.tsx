import React from 'react'

// Utils
import {
  render,
  screen,
  cleanup,
  fireEvent,
  RenderResult,
} from '@utils/testUtils'

// Components
import ReservationForm from '@components/ReservationForm'

describe('ReservationForm', () => {
  const userMock = {
    firstName: 'John',
    lastName: 'Wick',
    email: 'John@gmail.com',
    phone: '1231231231',
    date: '12/09/2022',
    time: '12:00 am',
    person: '2 Person',
  }

  const props = {
    onSubmitForm: jest.fn(),
    handleChangeDate: jest.fn(),
    onChangeField: jest.fn(),
    isDisable: false,
    isShowFullField: true,
  }

  let wrapper: RenderResult

  beforeEach(() => {
    wrapper = render(<ReservationForm {...props} />)
  })

  afterEach(() => {
    cleanup()
  })

  it('ReservationForm with full field matches snapshot', () => {
    const { container } = wrapper

    expect(container).toMatchSnapshot()
  })

  it('ReservationForm default matches snapshot', () => {
    const { container } = render(
      <ReservationForm {...props} isShowFullField={false} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('Should call onChange first name', () => {
    const firstName = screen.getByPlaceholderText('First Name')

    fireEvent.change(firstName, {
      target: { value: userMock.firstName },
    })

    expect(props.onChangeField).toBeCalled()
    expect(firstName).toHaveValue(userMock.firstName)
  })

  it('Should call onChange last name', () => {
    const lastName = screen.getByPlaceholderText('Last Name')

    fireEvent.change(lastName, {
      target: { value: userMock.lastName },
    })

    expect(props.onChangeField).toBeCalled()
    expect(lastName).toHaveValue(userMock.lastName)
  })

  it('Should call onChange email', () => {
    const email = screen.getByPlaceholderText('Email')

    fireEvent.change(email, {
      target: { value: userMock.email },
    })

    expect(props.onChangeField).toBeCalled()
    expect(email).toHaveValue(userMock.email)
  })

  it('Should call onChange phone', () => {
    const phone = screen.getByPlaceholderText('Phone')

    fireEvent.change(phone, {
      target: { value: userMock.phone },
    })

    expect(props.onChangeField).toBeCalled()
    expect(phone).toHaveValue(userMock.phone)
  })

  it('Should onchange date picker ', () => {
    const datePicker = screen.getByPlaceholderText('MM/DD/YYYY')

    fireEvent.change(datePicker, { target: { value: userMock.date } })

    expect(datePicker).toHaveValue(userMock.date)
  })

  it('Should onchange time ', () => {
    const time = screen.getByTestId('selectTime')

    fireEvent.change(time, { target: { value: userMock.time } })

    expect(time).toHaveValue(userMock.time)
  })

  it('Should onchange person ', () => {
    const person = screen.getByTestId('selectPerson')

    fireEvent.change(person, { target: { value: userMock.person } })

    expect(person).toHaveValue(userMock.person)
  })

  it('Should call submit form', () => {
    const button = screen.getByRole('button')

    fireEvent.submit(button)

    expect(props.onSubmitForm).toBeCalled()
  })
})
