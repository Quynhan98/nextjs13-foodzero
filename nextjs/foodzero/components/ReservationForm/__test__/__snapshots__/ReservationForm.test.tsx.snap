// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationForm ReservationForm default matches snapshot 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="chakra-form-control css-qjxwnb"
      role="group"
    >
      <div
        class="css-yjtq8f"
      >
        <h3
          class="chakra-heading css-17j5m7v"
        >
          Make a Reservation
        </h3>
        <p
          class="chakra-text css-1ce86fs"
        >
          Get in touch with restaurant
        </p>
      </div>
      <div
        class="css-4yycd6"
      >
        <div
          class="css-1payax8"
        >
          <div
            class="css-0"
          >
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class="chakra-input css-1sz8sj0"
                  id="field-:r2:"
                  placeholder="MM/DD/YYYY"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="css-0"
          >
            <div
              class="chakra-select__wrapper css-17qe6bp"
            >
              <select
                class="chakra-select css-lknr9b"
                data-testid="selectTime"
                id="field-:r2:"
                name="time"
                style="padding-left: 40px;"
              >
                <option
                  value="11:00 am"
                >
                  11:00 am
                </option>
                <option
                  value="12:00 am"
                >
                  12:00 am
                </option>
                <option
                  value="13:00 pm"
                >
                  13:00 pm
                </option>
                <option
                  value="14:00 pm"
                >
                  14:00 pm
                </option>
                <option
                  value="15:00 pm"
                >
                  15:00 pm
                </option>
                <option
                  value="16:00 pm"
                >
                  16:00 pm
                </option>
                <option
                  value="17:00 pm"
                >
                  17:00 pm
                </option>
                <option
                  value="18:00 pm"
                >
                  18:00 pm
                </option>
                <option
                  value="19:00 pm"
                >
                  19:00 pm
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-1rn5cz7"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon chakra-select__icon css-m73ylg"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-lknr9b"
            data-testid="selectPerson"
            id="field-:r2:"
            name="person"
            style="padding-left: 40px;"
          >
            <option
              value="1 Person"
            >
              1 Person
            </option>
            <option
              value="2 Person"
            >
              2 Person
            </option>
            <option
              value="3 Person"
            >
              3 Person
            </option>
            <option
              value="4 Person"
            >
              4 Person
            </option>
            <option
              value="5 Person"
            >
              5 Person
            </option>
            <option
              value="6 Person"
            >
              6 Person
            </option>
            <option
              value="7 Person"
            >
              7 Person
            </option>
            <option
              value="8 Person"
            >
              8 Person
            </option>
            <option
              value="9 Person"
            >
              9 Person
            </option>
            <option
              value="10 Person"
            >
              10 Person
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-1rn5cz7"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-select__icon css-m73ylg"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="css-1nx6xr5"
      >
        <button
          class="chakra-button css-156b4l8"
          type="submit"
        >
          Book Now
        </button>
      </div>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`ReservationForm ReservationForm with full field matches snapshot 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="chakra-form-control css-qjxwnb"
      role="group"
    >
      <div
        class="css-yjtq8f"
      >
        <h3
          class="chakra-heading css-17j5m7v"
        >
          Make a Reservation
        </h3>
        <p
          class="chakra-text css-1j19xza"
        >
          Get in touch with restaurant
        </p>
      </div>
      <div
        class="css-19580xy"
      >
        <input
          class="chakra-input css-sw9qjs"
          id="field-:r0:"
          name="firstName"
          placeholder="First Name"
          type="text"
          value=""
        />
        <input
          class="chakra-input css-sw9qjs"
          id="field-:r0:"
          name="lastName"
          placeholder="Last Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="css-1qnzkxj"
      >
        <input
          class="chakra-input css-1njo91c"
          id="field-:r0:"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="css-1qnzkxj"
      >
        <input
          class="chakra-input css-1njo91c"
          id="field-:r0:"
          name="phone"
          placeholder="Phone"
          type="tel"
          value=""
        />
      </div>
      <div
        class="css-r2rtav"
      >
        <div
          class="css-1payax8"
        >
          <div
            class="css-0"
          >
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class="chakra-input css-1cwrfhe"
                  id="field-:r0:"
                  placeholder="MM/DD/YYYY"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="css-0"
          >
            <div
              class="chakra-select__wrapper css-sp79o5"
            >
              <select
                class="chakra-select css-lknr9b"
                data-testid="selectTime"
                id="field-:r0:"
                name="time"
                style="padding-left: 40px;"
              >
                <option
                  value="11:00 am"
                >
                  11:00 am
                </option>
                <option
                  value="12:00 am"
                >
                  12:00 am
                </option>
                <option
                  value="13:00 pm"
                >
                  13:00 pm
                </option>
                <option
                  value="14:00 pm"
                >
                  14:00 pm
                </option>
                <option
                  value="15:00 pm"
                >
                  15:00 pm
                </option>
                <option
                  value="16:00 pm"
                >
                  16:00 pm
                </option>
                <option
                  value="17:00 pm"
                >
                  17:00 pm
                </option>
                <option
                  value="18:00 pm"
                >
                  18:00 pm
                </option>
                <option
                  value="19:00 pm"
                >
                  19:00 pm
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-1rn5cz7"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon chakra-select__icon css-m73ylg"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-lknr9b"
            data-testid="selectPerson"
            id="field-:r0:"
            name="person"
            style="padding-left: 40px;"
          >
            <option
              value="1 Person"
            >
              1 Person
            </option>
            <option
              value="2 Person"
            >
              2 Person
            </option>
            <option
              value="3 Person"
            >
              3 Person
            </option>
            <option
              value="4 Person"
            >
              4 Person
            </option>
            <option
              value="5 Person"
            >
              5 Person
            </option>
            <option
              value="6 Person"
            >
              6 Person
            </option>
            <option
              value="7 Person"
            >
              7 Person
            </option>
            <option
              value="8 Person"
            >
              8 Person
            </option>
            <option
              value="9 Person"
            >
              9 Person
            </option>
            <option
              value="10 Person"
            >
              10 Person
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-1rn5cz7"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-select__icon css-m73ylg"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="css-ffarb6"
      >
        <button
          class="chakra-button css-vjoqr8"
          type="submit"
        >
          Book Now
        </button>
      </div>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
